src/set_theory/zfc.lean

```diff
@@ -393,8 +375,7 @@ theorem ext_iff {x y : Set.{u}} : (∀ z : Set.{u}, z ∈ x ↔ z ∈ y) ↔ x =
 ⟨ext, λ h, by simp [h]⟩
 
 /-- The empty ZFC set -/
-def empty : Set := mk ∅
-instance : has_emptyc Set := ⟨empty⟩
+instance : has_emptyc Set := ⟨mk ∅⟩
```

Comment:
The purpose of the extra definition is so that we have a name for the thing that `∅` is notation for. Otherwise we can't refer to `Set.empty` directly; this is following the same pattern as `nat.add`, `int.mul`, `list.append` etc.

Reply:
We do have a name though: `has_emptyc.emptyc`.

There's many other places in `mathlib` where definitions are either privated or completely inlined into their instances, like `real.has_add`, `complex.has_add`, `set.has_insert`, `finsupp.has_zero`. One could argue that the reason your examples aren't inlined is because they use the equation compiler, which wouldn't work otherwise, though there's also examples of not inlined instances where the equation compiler isn't involved, like `finset.has_empty` or `multiset.has_empty`. We clearly don't have a fixed convention here.

I oppose having these extra definitions for two reasons:
- they mean one extra layer of unfolding: e.g. first `Set.has_emptyc`, then `Set.empty`.
- if someone accidentally tries using them directly, all the usual `simp` lemmas and `rw`s won't work.

That said, I don't really care enough to hold up this PR because of this. If you want, I'll revert the inlines.

Reply:
One thing though: if we are to keep the definitions, they should be protected. Both because they're common names that exist in other namespaces, and so that you're always using the instances unless you explicitly want to use the bare definition for... some reason. Especially `insert`, since the instance and the definition have the same name, which easily leads to the aforementioned `simp` and `rw` troubles.

Reply:
> We clearly don't have a fixed convention here.

When there is no fixed convention, the best thing to do is to not change things from one convention to another unless you have a good reason to (i.e. it is actually blocking something) or you are trying to implement a convention across the board (in which case there would be lots of other places to change, and a zulip discussion preceding the change). This kind of churn makes reviewing your PRs unnecessarily complicated.

Reply:
> One thing though: if we are to keep the definitions, they should be protected.

Fair enough. Most of the definitions in this section already are protected.

Reply:
I apologize then, I was under the false impression that these sorts of inlines were uncontroversial.